manager/manager.go

```diff
@@ -570,39 +515,30 @@ func (m *Manager) updateKEK(ctx context.Context, cluster *api.Cluster, r remotes
 			break
 		}
 	}
-	switch {
-	case subtle.ConstantTimeCompare(kek, m.keyRotator.GetCurrentKEK()) == 1:
-		// if the KEK hasn't changed, do nothing
-		return nil
-	case m.keyRotator.GetCurrentKEK() == nil:
-		// it was previously unencrypted, and now needs to be encrypted - trigger
-		// a full rotation of the TLS keys too
-		if err := ca.RenewTLSConfigNow(ctx, securityConfig, r); err != nil {
-			return err
-		}
-		// Now that we have a new TLS certificate, let's load it from disk and
-		// update our key rotator
-		headers, _, err := securityConfig.KeyReader().ReadHeaders()
-		if err != nil {
-			logger.WithError(err).Errorf("failed to read refreshed TLS key from disk")
-			return err
-		}
-		m.keyRotator.LoadFromHeaders(headers, kek)
-		// trigger a raft DEK rotation
-		m.keyRotator.RotationNotify() <- struct{}{}
-	default:
-		// the kek has just changed - all we just to do is re-encrypt the DEK and the cert
-		if err := securityConfig.KeyWriter().RotateKEK(kek); err != nil {
-			logger.WithError(err).Errorf("failed to re-encrypt TLS key")
-			return err
-		}
-		m.keyRotator.currentKEK = kek
-		logger.Debugf("KEK rotated successfully")
+	oldKEK, updated, err := m.keyRotator.MaybeUpdateKEK(kek)
+	if err != nil {
+		logger.WithError(err).Errorf("failed to re-encrypt TLS key")
+		return err
+	}
+	if updated {
+		logger.Debug("rotated KEK")
 	}
+	if updated && oldKEK == nil {
+		go func() {
+			// a best effort attempt to update the TLS certificate
+			if err := ca.RenewTLSConfigNow(ctx, securityConfig, r); err != nil {
+				logger.WithError(err).Errorf("failed to download new TLS certificate after locking the cluster")
+			}
+		}()
+	}
+	if _, err = m.keyRotator.MaybeUpdatePending(); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func (m *Manager) watchForKEKChanges(ctx context.Context) error {
+func (m *Manager) watchForKEKChanges(ctx context.Context, stopWatch, watchDone chan struct{}) error {
```

Comment:
Instead of having a `stopWatch` channel, you could cancel the context passed into this function.